### ISSUE TITLE
Create low-level generic mission AI in LUA

### ISSUE THREAD

**Issue by HenrytheSlav** [OpenRA/OpenRA#11187]
After discussion with @abcdefg30 I proposed creation of basic AI that will be supplies with missions if appropriate (base building/destroying missions).

One of posibilities to balance the original mission feeling and this is to create new "very hard" or rather "alternative" difficulty oriented on making missions more living at minor cost in authenticity. But it is alternative to pointless wiping out sitting duck AI we have currently.

This AI should be capable of:

- [ ] Base (re-)building
- [ ] Unit production
- [ ] Working eco (related to previous points)
- [ ] Harvester and refinery rebuilding and protection in case of common attacks/snipes.

I propose to use the buildup hack from Soviet 05 created by @abcdefg30 and supply maps with moderate reinforcements, linked to the eco. 

The missions might have to be altered as to keep them consistent with this AI and give them new "altered" dificulty strike teams/events etc. **Point of this is to make campaign challenging, create unorthodox situations and force player into quick response, while keeping the game MOSTLY fair** (you are playing against dumb AI trying to accomplish complicated tasks after all.

Please, feel free to add to discussion of this project and propose what else would you like to change, and how would you like to improve missions gameplay. And pls keep discussion civil :hamster:

**Comment by Biofreak1987** [comment]
I would really welcome that for porting campaigns. 
Right now I'm using a modified ai script from soviet 04 for all the missions I'm working. It would reduce the amount of duplicated code over a majority of campaign missions.

Features like automaticly rebuilding structures would be neat as well.

**Comment by HenrytheSlav** [comment]
Yes, I made soviet 04 too, mostly based on Soviet 05 in which I spent most time, learning lua and creating somehow entertaining mission (I hope). Only S05 has any buildup created AFAIK.

**Comment by GraionDilach** [comment]
One of the very few actual impressive feats of HackyAI is that it doesn't cheat. I'm not certain if allowing the AI to cheat is actually necessary.

**Comment by HenrytheSlav** [comment]
Graion, I don't plan to allow cheats for AI. I plan to give it advantage in form of reinforcements etc, which will give player something to be occupied with. HackyAI is for skirmishes where you want to have equal chances. We are talking on missins, where you are usually using small force to attack much bigger base (see Soviet10, Allied8) and the defence and AI assault on your possition should be correspondingly fierce.

**Comment by GraionDilach** [comment]
Ah, sorry then, I thought the bolded part meant cheating, not reinforcement triggers. Ignore me then.

**Comment by lovalmidas** [comment]
Hi,

I am relatively new to Lua scripting used in OpenRA, but I have been experimenting pieces of it to build a working AI that can provide a good degree of intelligence, or even **mimic player behavior**.

Some of my works are here. (Pardon me if you see code from elsewhere, especially missions)
http://resource.openra.net/maps/14893/
http://resource.openra.net/maps/15020/


My _ultimate_ goal is to build a set of Lua scripts that can perform the following tasks.
**Already existent**
- [ ] A1. Given building types and locations, perform base building.
- [ ] A2. Given factory types and a list of actors, perform unit production.
- [ ] A3. Holding lists for units - this can be used to transfer units to various sub-lists (a.k.a. teamtypes) for little tasks.
- [ ] A4. Stance changing ("no way I will continue to defend/hold if a V2RL keeps shooting me!")
- [ ] A5. Given waypoints, somewhat-working Allied and Soviet aircraft movements.

**Probably possible to build/refine.**
Economy
- [ ] B1a. Given cost and buildtime, trickle expenses (like how it is done for the player) every _n_ ticks.
- [ ] B1b. Given cost and buildtime, maintain a "progress" counter for each factory/production queue. (place on map/activate at 100%)
- [ ] B1c. If AI money is insufficient, put trickle expenses and "progress" counter on hold.
- [ ] B1d. Provide refunds and related customer service for AI.
- [ ] B1e. Player engineers can disrupt AI production, OpenRA style.
Teamtype and Actor behavior
- [ ] B2a. Extend _A3_ to include a sub-list for base defense.
- [ ] B2b. Extend _A3_ to include sub-lists for ore miner defense, that can be cleared when there are no threats.
- [ ] B2c. Offensive sub-lists to keep track of kills / losses / actor health.
- [ ] B2d. If losses are sustained, attacking sub-list to regroup and reinforce from holding list.
- [ ] B2e. Make a sub-list for repairing units.
- [ ] B2f. Extend _B2e_ to include heal/repair by medic/mechanic. 
- [ ] B2g. Extend _B2c_ + _B2d_ + location of enemy attacker to produce "formations". (lol micro)
Actor scripts
- [ ] B3a. Scripts: Attack infantry-only / Attack tanks-only / Attack structures-only (may be currently possible using rules.yaml, weapons.yaml and Lua scripts, "e7.noattacktanks")
- [ ] B3b. Dogs with a greater carnal desire to pursue thy enemy.
Base building enhancements
- [ ] B4a. AI to prioritize certain buildings (power/factory/refinery) in base building
- [ ] B4b. Given _B4a_, AI can note all its pre-placed buildings (with allowance for exceptions) and append them in its base building list.
- [ ] B4c. AI to prioritize buildings closer to base? Mostly aesthetic, but with _B4c_ this can be helpful.

**Probably not possible given the limited scope of Lua functions, yet**
**Require access to CanPowerDown or related controls**
- [ ] C1. AI to toggle power off when on low power (a hacky method exists: transfer ownership)
- [ ] C2. AI to toggle power on when on sufficient power.
- [ ] C3. AI to keep track of max power expenditure vs actual power usage, to help prioritize power plants.
- [ ] C4. AI to 'divert' power between base defenses based on surrounding threat.

**Require access to the terrain and overlay layers** 
- [ ] D1. AI to _choose_ building placement locations, obeying terrain limits and base adjacency.
- [ ] D2. AI to _choose_ its own attack waypoints based on terrain and enemy attack patterns
- [ ] D3. AI to scan for ore to prioritize refinery placements
- [ ] D4. Extend _D2_ + _D3_ for MCV expansion.
- [ ] D5. Factor base expansions into _D1_.
- [ ] D6. Extend _D4_ for other expansion styles (AI base push?)
- [ ] D7. Extend _D2_ for learning AI. (??!)


Probably should end my little rant somewhere.

**Comment by lovalmidas** [comment]
With http://resource.openra.net/maps/15220/, the following points should be fulfilled through Lua scripting.

B1a, B1b, B1c, B4a, B4b, B4c.
Also, unit and structure production should follow prerequisites as defined in the Lua script. It is kept separate from yaml prerequisites so you can tailor the AI build without influencing the player.

The scripts likely need optimization and cleaning, but if anyone wants to use them, feel free.
